zebra-state/src/service.rs

```diff
@@ -285,55 +287,36 @@ impl StateService {
         stop: Option<block::Hash>,
     ) -> Vec<block::Hash> {
         let mut res: Vec<block::Hash> = Vec::new();
-
-        // Known blocks can't be empty
-        if known_blocks.is_empty() {
+        // Get chain tip:
+        // We can get a block locator request before we have downloaded the genesis block,
+        // if this is the case we return an empty response.
+        let chain_tip = self.tip();
+        if chain_tip.is_none() {
             return res;
         }
-
-        // Get chain tip
-        let (chain_tip_height, ..) = self.tip().expect("tip must always be available");
-
-        // Get requested tip
-        let locator_tip_hash = known_blocks.first().expect("we must have a tip hash");
-        let locator_tip_height = self
-            .height_by_hash(*locator_tip_hash)
-            .expect("tip hash must have a height");
-        let chain = self.chain(*locator_tip_hash);
-
-        // Requested tip must be below our chain tip
-        if locator_tip_height >= chain_tip_height {
-            return res;
+        let chain_tip_height = chain_tip.unwrap().0;
+
+        // Get requested tip:
+        // Use the genesis hash if locator is empty or if no hash in the
+        // locator is present in our best chain.
+        let mut locator_tip_hash = genesis_hash(self.network);
+        let mut locator_tip_height = block::Height(1);
+        if !known_blocks.is_empty() {
+            for hash in known_blocks {
+                let height_hash = self.height_by_hash(hash);
+                if let Some(height_hash) = height_hash {
+                    locator_tip_hash = hash;
+                    locator_tip_height = height_hash;
+                    break;
+                }
+            }
         }
```

Comment:
```suggestion
        let (locator_tip_hash, locator_tip_height) = known_blocks
            .into_iter()
            .filter_map(|hash| self.height_by_hash(hash).map(|height| (hash, height)))
            .next()
            .unwrap_or_else(|| (genesis_hash(self.network), block::Height(1)));
```

Reply:
honestly tho, I'm not sure this is easier to read. @teor2345 and @oxarbitrage lmk what you think, feel free to ignore this suggestion if you prefer the old version.

Reply:
The code was a lot simpler once I split it up into separate functions, and tweaked the state interface a bit.

So this code no longer exists.